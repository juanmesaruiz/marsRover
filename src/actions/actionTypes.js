export const ROVER_CLEAR = 'ROVER@CLEAR';
export const ROVER_CLEAR_POSITION_LOG = 'ROVER@CLEAR_POSITION_LOG';
export const ROVER_DELETE_OBSTACLE = 'ROVER@DELETE_OBSTACLE';
export const ROVER_NEW_INSTRUCTIONS = 'ROVER@NEW_INSTRUCTIONS';
export const ROVER_NEW_MOVE = 'ROVER@NEW_MOVE';
export const ROVER_SET_GRID = 'ROVER@SET_GRID';
export const ROVER_SET_OBSTACLES = 'ROVER@SET_OBSTACLES';
export const ROVER_SET_POSITION = 'ROVER@SET_POSITION';
