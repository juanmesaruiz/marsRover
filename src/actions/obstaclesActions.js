import { ROVER_SET_OBSTACLES } from './actionTypes';

export const createObstacles = obstacles => ({
  type: ROVER_SET_OBSTACLES,
  payload: {
    obstacles,
  },
});
