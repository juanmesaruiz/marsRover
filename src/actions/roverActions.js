import { ROVER_CLEAR, ROVER_NEW_MOVE, ROVER_SET_POSITION } from './actionTypes';

export const roverClear = () => ({
  type: ROVER_CLEAR,
});

export const roverSetPosition = position => ({
  type: ROVER_SET_POSITION,
  payload: {
    position,
  },
});

export const roverNewKeyboardMove = ({ code, grid, obstaclesCoordinates }) => ({
  type: ROVER_NEW_MOVE,
  payload: {
    code,
    grid,
    obstaclesCoordinates,
  },
});
