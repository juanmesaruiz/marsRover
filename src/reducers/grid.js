import { ROVER_CLEAR } from '../actions/actionTypes';

const defaultState = { x: 50, y: 25 };

const grid = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    default:
      return state;
  }
};

export default grid;
