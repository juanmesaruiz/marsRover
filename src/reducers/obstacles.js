import { ROVER_CLEAR, ROVER_SET_OBSTACLES } from '../actions/actionTypes';

const defaultState = [];

const grid = (state = defaultState, action) => {
  switch (action.type) {
    case ROVER_CLEAR:
      return defaultState;

    case ROVER_SET_OBSTACLES: {
      const { obstacles } = action.payload;
      return [...state, ...obstacles];
    }

    default:
      return state;
  }
};

export default grid;
