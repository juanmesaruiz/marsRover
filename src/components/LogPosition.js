import React, { useCallback } from 'react';
import { connect } from 'react-redux';

import { roverClearPositionLog } from '../actions/roverActions';
import { getRover } from '../reducers';

import Button from './library/Button';

const LogPosition = ({
  rover: { current, direction, log },
  roverClearPositionLog,
}) => {
  const handleClearLog = useCallback(roverClearPositionLog, [
    roverClearPositionLog,
  ]);

  return (
    <>
      {current && (
        <p>
          Current position: X {current.x}, Y {current.y}
        </p>
      )}
      <p>Current direction: {direction}</p>
      <p>Last positions: </p>
      <Button onClick={handleClearLog}>Clear Log</Button>
      {log.map(({ x, y }, i) => (
        <p key={i}>
          X {x}, Y {y}
        </p>
      ))}
    </>
  );
};

const mapStateToProps = state => ({
  rover: getRover(state),
});

export default connect(mapStateToProps, { roverClearPositionLog })(LogPosition);
