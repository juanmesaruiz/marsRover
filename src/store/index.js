import { applyMiddleware, compose, createStore } from 'redux';
import thunk from 'redux-thunk';

import reducers from '../reducers';

const configureStore = () => {
  const initialState = {};
  const enhancers = [];
  const devToolsExtension = window.__REDUX_DEVTOOLS_EXTENSION__;

  if (typeof devToolsExtension === 'function') {
    enhancers.push(devToolsExtension({ name: '- ROVER MARS -' }));
  }

  const composedEnhancers = compose(applyMiddleware(thunk), ...enhancers);

  const store = createStore(reducers, initialState, composedEnhancers);

  return { store };
};

export default configureStore;
