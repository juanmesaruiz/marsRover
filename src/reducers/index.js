import { combineReducers } from 'redux';

import grid from './grid';
import obstacles from './obstacles';
import rover from './rover';

const reducers = {
  grid,
  obstacles,
  rover,
};

export const getGrid = state => state.grid;
export const getObstacles = state => state.obstacles;
export const getRover = state => state.rover;

export default combineReducers(reducers);
