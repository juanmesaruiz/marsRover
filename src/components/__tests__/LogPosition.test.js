import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';

import { roverClearPositionLog } from '../../actions/roverActions';
import { getRover } from '../../reducers';

import LogPosition from '../LogPosition';

import store from '../../store/__mocks__/mockStore';
import state from '../../store/__mocks__/mockState';

jest.mock('../../actions/roverActions');
jest.mock('../../reducers');

describe('LogPosition test suite', () => {
  const component = (
    <Provider store={store}>
      <LogPosition />
    </Provider>
  );

  const setup = () => render(component);

  beforeEach(() => {
    roverClearPositionLog.mockReturnValue({
      type: 'roverClearPositionLog',
    });
    getRover.mockReturnValue(state.rover);
  });

  afterEach(jest.clearAllMocks);

  it('Renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(component, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('Should match snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it('Should fire roverClearPositionLog action', () => {
    const { getByText } = setup();
    fireEvent.click(getByText('Clear Log'));

    expect(roverClearPositionLog).toHaveBeenCalledTimes(1);
  });
});
