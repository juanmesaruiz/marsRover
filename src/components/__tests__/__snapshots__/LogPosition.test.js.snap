// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogPosition test suite Should match snapshot 1`] = `
.c1 {
  cursor: pointer;
}

.c0 {
  display: inline-block;
  height: 376px;
  overflow: hidden;
  border: 1px solid black;
  margin-left: 30px;
  padding: 20px 20px;
  width: 200px;
}

.c0 .current {
  display: block;
  margin: 0 0 5px;
}

.c0 .direction {
  margin: 20px 0 18px;
}

.c0 .delete-button {
  float: right;
  margin-right: 30px;
}

.c0 .title {
  margin: 23px 0 0 0;
}

.c0 .positions-container {
  height: 225px;
  overflow: auto;
}

<div>
  <div
    class="c0"
  >
    <div>
      <span
        class="current"
      >
        Current position:
      </span>
      <b>
        X 
        2
        , Y 
        2
      </b>
    </div>
    <p
      class="direction"
    >
      Current direction: 
      <b>
        east
      </b>
    </p>
    <button
      class="c1"
    >
      Clear Log
    </button>
    <h5
      class="title"
    >
      Last positions: 
    </h5>
    <div
      class="positions-container"
    >
      <p>
        Obstacle - X 0, Y 3
      </p>
      <p>
        X 0, Y 1
      </p>
      <p>
        X 0, Y 0
      </p>
    </div>
  </div>
</div>
`;
